core/common/src/GenericInstanceFilter.ts

```diff
@@ -0,0 +1,187 @@
+  }
+  export type Values =
+    | string
+    | number
+    | boolean
+    | Date
+    | GenericInstanceFilterRuleValue.Point2d
+    | GenericInstanceFilterRuleValue.Point3d
+    | GenericInstanceFilterRuleValue.InstanceKey;
+}
+
+/**
+ * Defines single filter rule.
+ * @beta
+ */
+export interface GenericInstanceFilterRule {
+  /**
+   * Alias of the source to access this property. For related properties `sourceAlias` should match
+   * `GenericInstanceFilterRelatedInstanceDescription.alias` of one `GenericInstanceFilter.relatedInstances`.
+   */
+  sourceAlias: string;
+  /**
+   * Property name for accessing property value.
+   */
+  propertyName: string;
+  /**
+   * Comparison operator that should be used to compare property value.
+   */
+  operator: GenericInstanceFilterRuleOperator;
+  /**
+   * Value to which property values is compared to. For unary operators value is 'undefined'.
+   */
+  value?: GenericInstanceFilterRuleValue;
+  /**
+   * Type name of the property.
+   */
+  propertyTypeName: string;
+}
+
+/**
+ * Type definition that describes operators supported by `GenericInstanceFilterRuleGroup`.
+ * @beta
+ */
+export type GenericInstanceFilterRuleGroupOperator = "and" | "or";
+
+/**
+ * Group of filter rules joined by logical operator.
+ * @beta
+ */
+export interface GenericInstanceFilterRuleGroup {
+  /**
+   * Operator that should be used to join rules.
+   */
+  operator: GenericInstanceFilterRuleGroupOperator;
+  /**
+   * List of rules or rule groups that should be joined by `operator`.
+   */
+  rules: Array<GenericInstanceFilterRule | GenericInstanceFilterRuleGroup>;
+}
+
+/**
+ * Describes related instance whose property was used in the filter.
+ * @beta
+ */
+export interface GenericInstanceFilterRelatedInstanceDescription {
+  /**
+   * Describes path that should be used to reach related instance from the source.
+   */
+  path: GenericInstanceFilterRelationshipStep[];
+  /**
+   * Related instance alias. This alias match `sourceAlias` in all filter rules where
+   * properties of this related instance is used.
+   */
+  alias: string;
+}
+
+/**
+ * Describes single step between source class and target class.
+ * @beta
+ */
+export interface GenericInstanceFilterRelationshipStep {
+  /** Class name of the source class. */
+  sourceClassName: string;
+  /** Class name of the target class. */
+  targetClassName: string;
+  /** Class name of the relationship class that should be used to move from source to target. */
+  relationshipClassName: string;
+  /**
+   * Flag that describes if this step follows relationship class in `Forward` or `Backward` direction.
+   * If step follows relationship class in `Forward` direction than `sourceClassName` matches relationship source class and `targetClassName` matches relationship target class.
+   * Otherwise `sourceClassName` matches relationship target class and `targetClassName` matches relationship source class.
+   */
+  isForwardRelationship: boolean;
```

Comment:
I don't think "forward" and "backward" should be specified in backticks, since that's not an enum value or something like that.

```suggestion
  /**
   * A flag that describes if this step follows relationship class in forward or backward direction.
   * If the step follows relationship in forward direction then `sourceClassName` matches relationship's source class and `targetClassName` matches relationship's target class.
   * Otherwise, `sourceClassName` matches relationship's target class and `targetClassName` matches relationship's source class.
   */
  isForwardRelationship: boolean;
```